File: bot/models/execution.py
"""Canonical execution lifecycle persistence for Terminal money paths.

MONEY-PATH: these tables provide operational truth only. Economic truth belongs
in the separate ledger introduced by #908.
"""

from datetime import datetime
import uuid

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    JSON,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.sql import func

from database.db import Base


def _new_uuid() -> str:
    return str(uuid.uuid4())


class ExecutionIntent(Base):
    """User/agent economic intent before an execution plan is selected."""

    __tablename__ = "execution_intents"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=True, index=True)
    principal_key = Column(String(160), nullable=False, index=True)
    idempotency_key = Column(String(128), nullable=True)

    intent_type = Column(String(40), nullable=False)
    side = Column(String(12), nullable=True)
    from_chain = Column(String(50), nullable=True)
    to_chain = Column(String(50), nullable=True)
    from_asset = Column(String(128), nullable=True)
    to_asset = Column(String(128), nullable=True)
    requested_quantity = Column(String(78), nullable=True)
    requested_notional = Column(String(78), nullable=True)
    constraints_json = Column(JSON, nullable=True)
    metadata_json = Column(JSON, nullable=True)

    created_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime,
        default=datetime.utcnow,
        server_default=func.now(),
        onupdate=datetime.utcnow,
        nullable=False,
    )

    __table_args__ = (
        UniqueConstraint(
            "principal_key",
            "idempotency_key",
            name="uq_exec_intent_principal_idempotency",
        ),
    )


class ExecutionCandidatePlan(Base):
    """One feasible or rejected execution plan considered for an intent."""

    __tablename__ = "execution_candidate_plans"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    intent_id = Column(String(36), ForeignKey("execution_intents.id"), nullable=False, index=True)
    ordinal = Column(Integer, nullable=False)
    substrate = Column(String(40), nullable=False)
    provider = Column(String(80), nullable=True)
    strategy = Column(String(80), nullable=True)
    feasible = Column(Boolean, nullable=False, default=True, server_default=text("true"))
    rejection_code = Column(String(80), nullable=True)
    expected_to_amount = Column(String(78), nullable=True)
    expected_cost_bps = Column(String(78), nullable=True)
    expected_duration_ms = Column(Integer, nullable=True)
    plan_json = Column(JSON, nullable=True)
    cost_json = Column(JSON, nullable=True)
    selected = Column(Boolean, nullable=False, default=False, server_default=text("false"))
    quote_expires_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)

    __table_args__ = (
        UniqueConstraint("intent_id", "ordinal", name="uq_exec_candidate_intent_ordinal"),
        Index("ix_exec_candidate_intent_selected", "intent_id", "selected"),
    )


class ExecutionParentOrder(Base):
    """Lifecycle owner for one chosen execution plan."""

    __tablename__ = "execution_parent_orders"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    intent_id = Column(String(36), ForeignKey("execution_intents.id"), nullable=False, index=True)
    selected_candidate_id = Column(
        String(36), ForeignKey("execution_candidate_plans.id"), nullable=True, index=True
    )
    resubmission_of_parent_id = Column(
        String(36), ForeignKey("execution_parent_orders.id"), nullable=True, index=True
    )

    state = Column(String(24), nullable=False, default="draft", server_default=text("'draft'"))
    state_version = Column(Integer, nullable=False, default=0, server_default=text("0"))
    strategy = Column(String(80), nullable=True)
    authorization_method = Column(String(40), nullable=True)
    requested_quantity = Column(String(78), nullable=True)
    filled_quantity = Column(String(78), nullable=False, default="0", server_default=text("'0'"))
    average_fill_price = Column(String(78), nullable=True)

    submit_idempotency_key = Column(String(128), nullable=True)
    request_fingerprint = Column(String(128), nullable=True)

    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime,
        default=datetime.utcnow,
        server_default=func.now(),
        onupdate=datetime.utcnow,
        nullable=False,
    )

    __table_args__ = (
        UniqueConstraint(
            "intent_id",
            "submit_idempotency_key",
            name="uq_exec_parent_intent_submit_key",
        ),
        Index("ix_exec_parent_state_updated", "state", "updated_at"),
    )


class ExecutionChildPlacement(Base):
    """One external order, transaction, or solver child submitted by a parent."""

    __tablename__ = "execution_child_placements"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    parent_order_id = Column(
        String(36), ForeignKey("execution_parent_orders.id"), nullable=False, index=True
    )
    child_sequence = Column(Integer, nullable=False)
    substrate = Column(String(40), nullable=False)
    provider = Column(String(80), nullable=True)
    venue = Column(String(80), nullable=True)
    chain = Column(String(50), nullable=True)
    side = Column(String(12), nullable=True)
    requested_quantity = Column(String(78), nullable=True)
    limit_price = Column(String(78), nullable=True)
    state = Column(String(32), nullable=False, default="created", server_default=text("'created'"))

    idempotency_key = Column(String(128), nullable=True)
    request_fingerprint = Column(String(128), nullable=True)
    external_order_id = Column(String(255), nullable=True)
    external_tx_hash = Column(String(255), nullable=True)
    external_intent_id = Column(String(255), nullable=True)

    submitted_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime,
        default=datetime.utcnow,
        server_default=func.now(),
        onupdate=datetime.utcnow,
        nullable=False,
    )

    __table_args__ = (
        UniqueConstraint(
            "parent_order_id",
            "child_sequence",
            name="uq_exec_child_parent_sequence",
        ),
        UniqueConstraint("provider", "external_order_id", name="uq_exec_child_provider_order"),
        Index("ix_exec_child_parent_state", "parent_order_id", "state"),
    )


class ExecutionFill(Base):
    """Authoritative external fill observed for a parent/child."""

    __tablename__ = "execution_fills"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    parent_order_id = Column(
        String(36), ForeignKey("execution_parent_orders.id"), nullable=False, index=True
    )
    child_placement_id = Column(
        String(36), ForeignKey("execution_child_placements.id"), nullable=True, index=True
    )
    external_source = Column(String(80), nullable=False)
    external_fill_id = Column(String(255), nullable=True)
    quantity = Column(String(78), nullable=False)
    price = Column(String(78), nullable=False)
    fee_amount = Column(String(78), nullable=True)
    fee_asset = Column(String(128), nullable=True)
    liquidity_role = Column(String(16), nullable=True)
    metadata_json = Column(JSON, nullable=True)
    occurred_at = Column(DateTime, nullable=False)
    observed_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)

    __table_args__ = (
        UniqueConstraint(
            "external_source",
            "external_fill_id",
            name="uq_exec_fill_source_external",
        ),
        Index("ix_exec_fill_parent_occurred", "parent_order_id", "occurred_at"),
    )


class ExecutionSettlement(Base):
    """Confirmation/finality/recovery state associated with execution value movement."""

    __tablename__ = "execution_settlements"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    parent_order_id = Column(
        String(36), ForeignKey("execution_parent_orders.id"), nullable=False, index=True
    )
    child_placement_id = Column(
        String(36), ForeignKey("execution_child_placements.id"), nullable=True, index=True
    )
    settlement_type = Column(String(40), nullable=False)
    external_source = Column(String(80), nullable=False)
    external_ref = Column(String(255), nullable=False)
    state = Column(String(32), nullable=False, default="pending", server_default=text("'pending'"))
    chain = Column(String(50), nullable=True)
    asset = Column(String(128), nullable=True)
    amount = Column(String(78), nullable=True)
    confirmations = Column(Integer, nullable=True)
    finality_target = Column(Integer, nullable=True)
    recovery_json = Column(JSON, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime,
        default=datetime.utcnow,
        server_default=func.now(),
        onupdate=datetime.utcnow,
        nullable=False,
    )

    __table_args__ = (
        UniqueConstraint(
            "external_source",
            "external_ref",
            "settlement_type",
            name="uq_exec_settlement_external",
        ),
        Index("ix_exec_settlement_parent_state", "parent_order_id", "state"),
    )


class ExecutionEvent(Base):
    """Immutable event used to replay one parent order's operational state."""

    __tablename__ = "execution_events"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    parent_order_id = Column(
        String(36), ForeignKey("execution_parent_orders.id"), nullable=False, index=True
    )
    sequence = Column(Integer, nullable=False)
    event_type = Column(String(64), nullable=False)
    from_state = Column(String(24), nullable=True)
    to_state = Column(String(24), nullable=True)
    payload_json = Column(JSON, nullable=True)
    actor_type = Column(String(24), nullable=True)
    actor_id = Column(String(160), nullable=True)
    correlation_id = Column(String(128), nullable=True, index=True)
    causation_id = Column(String(128), nullable=True)
    occurred_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)

    __table_args__ = (
        UniqueConstraint("parent_order_id", "sequence", name="uq_exec_event_parent_sequence"),
        Index("ix_exec_event_parent_occurred", "parent_order_id", "occurred_at"),
    )


class ExecutionOutbox(Base):
    """Transactional outbox row for at-least-once lifecycle event publication."""

    __tablename__ = "execution_outbox"

    id = Column(String(36), primary_key=True, default=_new_uuid)
    event_id = Column(String(36), ForeignKey("execution_events.id"), nullable=False)
    topic = Column(String(128), nullable=False)
    payload_json = Column(JSON, nullable=True)
    attempts = Column(Integer, nullable=False, default=0, server_default=text("0"))
    published_at = Column(DateTime, nullable=True)
    next_attempt_at = Column(DateTime, nullable=True)
    last_error = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, server_default=func.now(), nullable=False)

    __table_args__ = (
        UniqueConstraint("event_id", name="uq_execution_outbox_event_id"),
        Index("ix_exec_outbox_publish", "published_at", "next_attempt_at"),
    )

File: api-ts/src/db/schema/execution.ts
import {
	boolean,
	index,
	integer,
	jsonb,
	pgTable,
	text,
	timestamp,
	uniqueIndex,
	varchar,
} from 'drizzle-orm/pg-core'
import { users } from './users'

export const executionIntents = pgTable(
	'execution_intents',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		userId: integer('user_id').references(() => users.id),
		principalKey: varchar('principal_key', { length: 160 }).notNull(),
		idempotencyKey: varchar('idempotency_key', { length: 128 }),
		intentType: varchar('intent_type', { length: 40 }).notNull(),
		side: varchar('side', { length: 12 }),
		fromChain: varchar('from_chain', { length: 50 }),
		toChain: varchar('to_chain', { length: 50 }),
		fromAsset: varchar('from_asset', { length: 128 }),
		toAsset: varchar('to_asset', { length: 128 }),
		requestedQuantity: varchar('requested_quantity', { length: 78 }),
		requestedNotional: varchar('requested_notional', { length: 78 }),
		constraintsJson: jsonb('constraints_json'),
		metadataJson: jsonb('metadata_json'),
		createdAt: timestamp('created_at').defaultNow().notNull(),
		updatedAt: timestamp('updated_at').defaultNow().notNull(),
	},
	(table) => ({
		userIdx: index('ix_execution_intents_user_id').on(table.userId),
		principalIdx: index('ix_execution_intents_principal_key').on(table.principalKey),
		principalIdempotency: uniqueIndex('uq_exec_intent_principal_idempotency').on(
			table.principalKey,
			table.idempotencyKey,
		),
	}),
)

export const executionCandidatePlans = pgTable(
	'execution_candidate_plans',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		intentId: varchar('intent_id', { length: 36 })
			.notNull()
			.references(() => executionIntents.id),
		ordinal: integer('ordinal').notNull(),
		substrate: varchar('substrate', { length: 40 }).notNull(),
		provider: varchar('provider', { length: 80 }),
		strategy: varchar('strategy', { length: 80 }),
		feasible: boolean('feasible').default(true).notNull(),
		rejectionCode: varchar('rejection_code', { length: 80 }),
		expectedToAmount: varchar('expected_to_amount', { length: 78 }),
		expectedCostBps: varchar('expected_cost_bps', { length: 78 }),
		expectedDurationMs: integer('expected_duration_ms'),
		planJson: jsonb('plan_json'),
		costJson: jsonb('cost_json'),
		selected: boolean('selected').default(false).notNull(),
		quoteExpiresAt: timestamp('quote_expires_at'),
		createdAt: timestamp('created_at').defaultNow().notNull(),
	},
	(table) => ({
		intentIdx: index('ix_execution_candidate_plans_intent_id').on(table.intentId),
		intentOrdinal: uniqueIndex('uq_exec_candidate_intent_ordinal').on(
			table.intentId,
			table.ordinal,
		),
		selectedIdx: index('ix_exec_candidate_intent_selected').on(table.intentId, table.selected),
	}),
)

export const executionParentOrders = pgTable(
	'execution_parent_orders',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		intentId: varchar('intent_id', { length: 36 })
			.notNull()
			.references(() => executionIntents.id),
		selectedCandidateId: varchar('selected_candidate_id', { length: 36 }).references(
			() => executionCandidatePlans.id,
		),
		// SQLAlchemy owns the self-FK. Drizzle mirrors the column without a
		// recursive reference to keep the schema module initialization simple.
		resubmissionOfParentId: varchar('resubmission_of_parent_id', { length: 36 }),
		state: varchar('state', { length: 24 }).default('draft').notNull(),
		stateVersion: integer('state_version').default(0).notNull(),
		strategy: varchar('strategy', { length: 80 }),
		authorizationMethod: varchar('authorization_method', { length: 40 }),
		requestedQuantity: varchar('requested_quantity', { length: 78 }),
		filledQuantity: varchar('filled_quantity', { length: 78 }).default('0').notNull(),
		averageFillPrice: varchar('average_fill_price', { length: 78 }),
		submitIdempotencyKey: varchar('submit_idempotency_key', { length: 128 }),
		requestFingerprint: varchar('request_fingerprint', { length: 128 }),
		startedAt: timestamp('started_at'),
		completedAt: timestamp('completed_at'),
		createdAt: timestamp('created_at').defaultNow().notNull(),
		updatedAt: timestamp('updated_at').defaultNow().notNull(),
	},
	(table) => ({
		intentIdx: index('ix_execution_parent_orders_intent_id').on(table.intentId),
		candidateIdx: index('ix_execution_parent_orders_selected_candidate_id').on(
			table.selectedCandidateId,
		),
		resubmissionIdx: index('ix_execution_parent_orders_resubmission_of_parent_id').on(
			table.resubmissionOfParentId,
		),
		intentSubmitKey: uniqueIndex('uq_exec_parent_intent_submit_key').on(
			table.intentId,
			table.submitIdempotencyKey,
		),
		stateUpdatedIdx: index('ix_exec_parent_state_updated').on(table.state, table.updatedAt),
	}),
)

export const executionChildPlacements = pgTable(
	'execution_child_placements',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		parentOrderId: varchar('parent_order_id', { length: 36 })
			.notNull()
			.references(() => executionParentOrders.id),
		childSequence: integer('child_sequence').notNull(),
		substrate: varchar('substrate', { length: 40 }).notNull(),
		provider: varchar('provider', { length: 80 }),
		venue: varchar('venue', { length: 80 }),
		chain: varchar('chain', { length: 50 }),
		side: varchar('side', { length: 12 }),
		requestedQuantity: varchar('requested_quantity', { length: 78 }),
		limitPrice: varchar('limit_price', { length: 78 }),
		state: varchar('state', { length: 32 }).default('created').notNull(),
		idempotencyKey: varchar('idempotency_key', { length: 128 }),
		requestFingerprint: varchar('request_fingerprint', { length: 128 }),
		externalOrderId: varchar('external_order_id', { length: 255 }),
		externalTxHash: varchar('external_tx_hash', { length: 255 }),
		externalIntentId: varchar('external_intent_id', { length: 255 }),
		submittedAt: timestamp('submitted_at'),
		createdAt: timestamp('created_at').defaultNow().notNull(),
		updatedAt: timestamp('updated_at').defaultNow().notNull(),
	},
	(table) => ({
		parentIdx: index('ix_execution_child_placements_parent_order_id').on(table.parentOrderId),
		parentSequence: uniqueIndex('uq_exec_child_parent_sequence').on(
			table.parentOrderId,
			table.childSequence,
		),
		providerOrder: uniqueIndex('uq_exec_child_provider_order').on(
			table.provider,
			table.externalOrderId,
		),
		parentStateIdx: index('ix_exec_child_parent_state').on(table.parentOrderId, table.state),
	}),
)

export const executionFills = pgTable(
	'execution_fills',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		parentOrderId: varchar('parent_order_id', { length: 36 })
			.notNull()
			.references(() => executionParentOrders.id),
		childPlacementId: varchar('child_placement_id', { length: 36 }).references(
			() => executionChildPlacements.id,
		),
		externalSource: varchar('external_source', { length: 80 }).notNull(),
		externalFillId: varchar('external_fill_id', { length: 255 }),
		quantity: varchar('quantity', { length: 78 }).notNull(),
		price: varchar('price', { length: 78 }).notNull(),
		feeAmount: varchar('fee_amount', { length: 78 }),
		feeAsset: varchar('fee_asset', { length: 128 }),
		liquidityRole: varchar('liquidity_role', { length: 16 }),
		metadataJson: jsonb('metadata_json'),
		occurredAt: timestamp('occurred_at').notNull(),
		observedAt: timestamp('observed_at').defaultNow().notNull(),
	},
	(table) => ({
		parentIdx: index('ix_execution_fills_parent_order_id').on(table.parentOrderId),
		childIdx: index('ix_execution_fills_child_placement_id').on(table.childPlacementId),
		sourceFill: uniqueIndex('uq_exec_fill_source_external').on(
			table.externalSource,
			table.externalFillId,
		),
		parentOccurredIdx: index('ix_exec_fill_parent_occurred').on(
			table.parentOrderId,
			table.occurredAt,
		),
	}),
)

export const executionSettlements = pgTable(
	'execution_settlements',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		parentOrderId: varchar('parent_order_id', { length: 36 })
			.notNull()
			.references(() => executionParentOrders.id),
		childPlacementId: varchar('child_placement_id', { length: 36 }).references(
			() => executionChildPlacements.id,
		),
		settlementType: varchar('settlement_type', { length: 40 }).notNull(),
		externalSource: varchar('external_source', { length: 80 }).notNull(),
		externalRef: varchar('external_ref', { length: 255 }).notNull(),
		state: varchar('state', { length: 32 }).default('pending').notNull(),
		chain: varchar('chain', { length: 50 }),
		asset: varchar('asset', { length: 128 }),
		amount: varchar('amount', { length: 78 }),
		confirmations: integer('confirmations'),
		finalityTarget: integer('finality_target'),
		recoveryJson: jsonb('recovery_json'),
		createdAt: timestamp('created_at').defaultNow().notNull(),
		updatedAt: timestamp('updated_at').defaultNow().notNull(),
	},
	(table) => ({
		parentIdx: index('ix_execution_settlements_parent_order_id').on(table.parentOrderId),
		childIdx: index('ix_execution_settlements_child_placement_id').on(table.childPlacementId),
		externalKey: uniqueIndex('uq_exec_settlement_external').on(
			table.externalSource,
			table.externalRef,
			table.settlementType,
		),
		parentStateIdx: index('ix_exec_settlement_parent_state').on(
			table.parentOrderId,
			table.state,
		),
	}),
)

export const executionEvents = pgTable(
	'execution_events',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		parentOrderId: varchar('parent_order_id', { length: 36 })
			.notNull()
			.references(() => executionParentOrders.id),
		sequence: integer('sequence').notNull(),
		eventType: varchar('event_type', { length: 64 }).notNull(),
		fromState: varchar('from_state', { length: 24 }),
		toState: varchar('to_state', { length: 24 }),
		payloadJson: jsonb('payload_json'),
		actorType: varchar('actor_type', { length: 24 }),
		actorId: varchar('actor_id', { length: 160 }),
		correlationId: varchar('correlation_id', { length: 128 }),
		causationId: varchar('causation_id', { length: 128 }),
		occurredAt: timestamp('occurred_at').defaultNow().notNull(),
	},
	(table) => ({
		parentIdx: index('ix_execution_events_parent_order_id').on(table.parentOrderId),
		correlationIdx: index('ix_execution_events_correlation_id').on(table.correlationId),
		parentSequence: uniqueIndex('uq_exec_event_parent_sequence').on(
			table.parentOrderId,
			table.sequence,
		),
		parentOccurredIdx: index('ix_exec_event_parent_occurred').on(
			table.parentOrderId,
			table.occurredAt,
		),
	}),
)

export const executionOutbox = pgTable(
	'execution_outbox',
	{
		id: varchar('id', { length: 36 }).primaryKey(),
		eventId: varchar('event_id', { length: 36 })
			.notNull()
			.references(() => executionEvents.id),
		topic: varchar('topic', { length: 128 }).notNull(),
		payloadJson: jsonb('payload_json'),
		attempts: integer('attempts').default(0).notNull(),
		publishedAt: timestamp('published_at'),
		nextAttemptAt: timestamp('next_attempt_at'),
		lastError: text('last_error'),
		createdAt: timestamp('created_at').defaultNow().notNull(),
	},
	(table) => ({
		eventUnique: uniqueIndex('uq_execution_outbox_event_id').on(table.eventId),
		publishIdx: index('ix_exec_outbox_publish').on(table.publishedAt, table.nextAttemptAt),
	}),
)

export type ExecutionIntent = typeof executionIntents.$inferSelect
export type NewExecutionIntent = typeof executionIntents.$inferInsert
export type ExecutionCandidatePlan = typeof executionCandidatePlans.$inferSelect
export type ExecutionParentOrder = typeof executionParentOrders.$inferSelect
export type ExecutionChildPlacement = typeof executionChildPlacements.$inferSelect
export type ExecutionFill = typeof executionFills.$inferSelect
export type ExecutionSettlement = typeof executionSettlements.$inferSelect
export type ExecutionEvent = typeof executionEvents.$inferSelect
